R/convert_output_type.R

```diff
@@ -0,0 +1,280 @@
+    req_to_otid_cols <- "output_type_id"
+    if (!all(req_to_otid_cols %in% names(to_otid))) {
+      cli::cli_abort(c(
+        "the {.val {to_output_type}} element of {.arg to} did not contain
+        the required column {.val {req_to_otid_cols}}"
+      ))
+    }
+
+    # Validate columns for joining, if applicable
+    validate_join_by_cols(to_output_type, to_otid, model_out_tbl, return_missing_combos = FALSE)
+
+    if (to_output_type == "quantile") {
+      if (!is.numeric(to_otid$output_type_id)) {
+        cli::cli_abort(c(
+          "!" = "Values in {.arg to} representing {.val quantile} output type IDs
+          should be numeric"
+        ))
+      }
+      less_than_0 <- to_otid$output_type_id < 0
+      more_than_1 <- to_otid$output_type_id > 1
+      if (any(less_than_0) || any(more_than_1)) {
+        cli::cli_abort(c(
+          "!" = "Values in {.arg to} representing {.val quantile} output type IDs
+          should be between {.val {0L}} and {.val {1L}}.",
+          x = "Value{?s} outside this range found:
+          {.val {to_otid$output_type_id[less_than_0 | more_than_1]}}"
+        ))
+      }
+    }
+  }
+  to_otid
+}
+
+
+#' Validate the to join by columns for a single to output type
+#' @noRd
+validate_join_by_cols <- function(to_output_type, to_otid, model_out_tbl, return_missing_combos = FALSE) {
+  # joining columns must be part of task_id_cols
+  req_to_otid_cols <- "output_type_id"
+  task_id_cols <- subset_task_id_names(colnames(model_out_tbl))
+  join_by_cols <- names(to_otid)[!(names(to_otid) %in% req_to_otid_cols)]
+  invalid_cols <- join_by_cols[!(join_by_cols %in% task_id_cols)]
+  if (length(invalid_cols) > 0L) {
+    cli::cli_abort(c(
+      "x" = "the {.val to_output_type} element of {.arg to} included
+            {length(invalid_cols)} task ID{?s} that {?was/were} not present in
+            {.arg model_out_tbl}: {.val {invalid_cols}}"
+    ))
+  }
+
+  if (length(join_by_cols) > 0) {
+    for (join_col in join_by_cols) {
+      # joining columns must be of the same type
+      # `join_by_cols` type in `to` is coerced to that in `model_out_tbl`
+      class(to_otid[[join_col]]) <- class(model_out_tbl[[join_col]])
+    }
+
+    # task ID var values in to_otid must appear in associated model_out_tbl cols
+    # warn for the opposite case
+    missing_model_out_case <- dplyr::anti_join(to_otid, model_out_tbl, by = join_by_cols)
+    missing_to_case <- dplyr::anti_join(model_out_tbl, to_otid, by = join_by_cols)
+    if (nrow(missing_model_out_case) > 0) {
+      if (!return_missing_combos) {
+        cli::cli_warn(c(
+          x = "Some task ID variable combos present in the {.val {to_output_type}} 
+            element of {.arg to} are missing from {.arg model_out_tbl}",
+          i = "Run {.arg validate_join_by_cols(to_output_type, to_otid, 
+            model_out_tbl, return_missing_combos = TRUE)} to see the missing cases"
+        ))
+      } else {
+        missing_model_out_case
+      }
```

Comment:
Suggestion: this control structure is identical to the one below. It might be good to abstract it into a function. 

```r
if (nrow(missing_model_out_case) > 0) {
  return_or_err(
    missing_model_out_case, 
    return_missing_combos = return_missing_combos, 
    to_output_type = to_output_type, 
    from_model_out = TRUE
  )
} else {
  return_or_err(
    missing_model_out_case, 
    return_missing_combos = return_missing_combos, 
    to_output_type = to_output_type, 
    from_model_out = TRUE
  )
}
```